addons/l10n_id/models/account_move.py

```diff
@@ -0,0 +1,152 @@
+
+        :return: a list with the format:
+            {
+                invoice: {
+                    'paid': True,
+                    'qr_statuses': {}
+                },
+                invoice: {
+                    'paid': False
+                }
+            }
+        """
+        result = {}
+        for invoice in self:
+            paid = False
+            data = []
+            # Looping to make requests is far from ideal, but we have no choices as they don't allow getting multiple QR result at once.
+            # Ensure to loop in reverse and check from the most recent QR code.
+            for qr_invoice in reversed(invoice.l10n_id_qris_invoice_details):
+                status_response = invoice.partner_bank_id._qris_fetch_status(qr_invoice)
+                data.append(status_response['data'])
+                if status_response['data'].get('qris_status') == 'paid':
+                    paid = True
+                    break  # We don't really need any unpaid data, as long as we know which QR is paid
+            result[invoice] = {
+                'paid': paid,
+                'qr_statuses': data,
+            }
+        return result
+
+    @api.model
+    def _process_unpaid_invoices(self, unpaid_statuses):
+        """
+        For unpaid invoices, the statuses received do not matter. We just check the QR dates and discard expired ones.
+        """
+        if not unpaid_statuses:
+            return
+
+        for invoice, invoice_information in unpaid_statuses.items():
+            qris_data_to_recheck = []
+            for qr_invoice in reversed(invoice.l10n_id_qris_invoice_details):
+                # The QR date is ib Jakarta time, so we ensure that we use correct timezones.
+                qris_datetime = fields.Datetime.to_datetime(
+                    qr_invoice['qris_creation_datetime']
+                ).replace(tzinfo=pytz.timezone('Asia/Jakarta'))
+                jakarta_now = fields.Datetime.context_timestamp(self.with_context(tz='Asia/Jakarta'), fields.Datetime.now())
+                # We will only reverify QR codes that have less than 30m of age, as after that they are no longer valid and will never be paid.
+                if (jakarta_now - qris_datetime).total_seconds() < 1800:
+                    qris_data_to_recheck.append(qr_invoice)
+            invoice.l10n_id_qris_invoice_details = qris_data_to_recheck
+
+    @api.model
+    def _process_paid_invoices(self, paid_statuses):
+        """
+        For unpaid invoices, the statuses received do not matter. We just check the QR dates and discard expired ones.
+        """
+        if not paid_statuses:
+            return
+
+        invoices = self.env['account.move']
+        for invoice, invoice_information in paid_statuses.items():
+            invoices |= invoice  # gather all invoices in a recordset for later.
+            # Find the paid QR to gather the payment detail
+            for qr_status in invoice_information['qr_statuses']:
```

Comment:
Shouldn't we have only one status when it is paid?

Reply:
It should be yes, but anyway it's better to keep the data types consistent and return a list in both cases.
(It's been rewritten a bit anyway)